File: workspace/solutions/go/s0/s3/s0/s3/s_0303_00/s_0303_00.go
/*
 * @lc app=leetcode.cn id=303 lang=golang
 *
 * [303] 区域和检索 - 数组不可变
 */

package s_0303_00

// @lc code=start
type NumArray struct {
	PrefixSums []int
}

func Constructor(nums []int) NumArray {
	prefixSums := make([]int, len(nums)+1)
	for i, v := range nums {
		prefixSums[i+1] = prefixSums[i] + v
	}
	return NumArray{
		PrefixSums: prefixSums,
	}
}

func (this *NumArray) SumRange(left int, right int) int {
	return this.PrefixSums[right+1] - this.PrefixSums[left]
}

/**
 * Your NumArray object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.SumRange(left,right);
 */
// @lc code=end

func (this NumArray) Func(left int, right int) int {
	return this.SumRange(left, right)
}

var Solution = Constructor

File: workspace/solutions/ecmascript/src/s0/s3/s0/s3/s_0303_00.js
/*
* @lc app=leetcode.cn id=303 lang=javascript
*
* [303] 区域和检索 - 数组不可变
*/

// @lc code=start
/**
 * @param {number[]} nums
 * @constructor
 */
function NumArray(nums) {
    this.prefix_sums = nums.slice();
    this.prefix_sums.reduce((sum, value, index, sums) => {
        sum += value;
        sums[index] = sum;
        return sum;
    }, 0);
};

/**
 * @param {number} left
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function (left, right) {
    return (this.prefix_sums[right] ?? 0) - (this.prefix_sums[left - 1] ?? 0);
};

/**
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */
// @lc code=end

NumArray.prototype.func = NumArray.prototype.sumRange;
export default NumArray;
